src/transactions/InvokeHostFunctionOpFrame.cpp

```diff
@@ -301,6 +279,103 @@ struct HostFunctionMetrics
     }
 };
 
+InvokeHostFunctionOpFrame::InvokeHostFunctionOpFrame(Operation const& op,
+                                                     OperationResult& res,
+                                                     TransactionFrame& parentTx)
+    : OperationFrame(op, res, parentTx)
+    , mInvokeHostFunction(mOperation.body.invokeHostFunctionOp())
+{
+}
+
+bool
+InvokeHostFunctionOpFrame::isOpSupported(LedgerHeader const& header) const
+{
+    return header.ledgerVersion >= 20;
+}
+
+bool
+InvokeHostFunctionOpFrame::doApply(AbstractLedgerTxn& ltx)
+{
+    throw std::runtime_error(
+        "InvokeHostFunctionOpFrame::doApply needs Config and base PRNG seed");
+}
+
+void
+InvokeHostFunctionOpFrame::maybePopulateDiagnosticEvents(
+    Config const& cfg, InvokeHostFunctionOutput const& output,
+    HostFunctionMetrics const& metrics)
+{
+    if (cfg.ENABLE_SOROBAN_DIAGNOSTIC_EVENTS)
+    {
+        xdr::xvector<DiagnosticEvent> diagnosticEvents;
+        for (auto const& e : output.diagnostic_events)
+        {
+            DiagnosticEvent evt;
+            xdr::xdr_from_opaque(e.data, evt);
+            diagnosticEvents.emplace_back(evt);
+        }
+
+        // add an additional diagnostic event that contains all the metrics
+        DiagnosticEvent de;
+        de.inSuccessfulContractCall = metrics.mSuccess;
+        de.event.type = ContractEventType::DIAGNOSTIC;
+        SCVec topics;
+        SCVal data(SCV_VEC);
+
+        topics = {
```

Comment:
Setting up the data like this doesn't really make it easy to parse. I think using one topic like "core_metrics" and a symbol->U64 Map in the data would be easier to consume.